R/module_treeview.R

```diff
@@ -0,0 +1,85 @@
+#' Shiny module for presenting the treeview data
+#'
+#' @param   id   Shiny/HTML ID for the module.
+#' @return   fluidRow UI elements for the module.
+
+treeviewUI = function(id) {
+  ns = shiny::NS(id)
+
+  shiny::fluidRow(
+    shiny::column(12,
+                  # choose type of treeview
+                  shiny::radioButtons(
+                    inputId = ns("widgetChoice"),
+                    label = "Select treeview",
+                    choices = c(
+                      "Logistic growth rate",
+                      "Simple logistic growth rate",
+                      "Simple trait log odds"
+                    ),
+                    inline = TRUE
+                  ),
+
+                  # show treeview widget
+                  shiny::wellPanel(
+                    ggiraph::girafeOutput(ns("treeview")),
+                    style = "background: white",
+                  ),
+                  shiny::br()
+    )
+  ) # end fluid row
+}
+
+#' Shiny server function for presenting the treeview data
+#'
+#' @inheritParams   treeviewUI
+#'
+#' @return   Creates a \code{moduleServer} that returns a reactive containing the cluster ID for
+#'   whichever node the user most recently hovered on.
+
+treeviewServer = function(id) {
+  shiny::moduleServer(
+    id,
+    function(input, output, session) {
+      # get selected cluster ID based on widget choice
+      selected_cluster_id = shiny::reactive({
+        get_selected_cluster_id(widgetChoice = input$widgetChoice,
+                                treeviewSelected = input$treeview_selected)
+      })
+
+      imported_ggtree = shiny::reactive({
+        filename = get_filename(input$widgetChoice)
+        readRDS(filename)
+      })
+
+      # create plotly output from saved ggplot2 outputs
+      output$treeview = ggiraph::renderGirafe({
+        tooltip_css = paste0(
+          "background-color:black;",
+          "color:grey;",
+          "padding:14px;",
+          "border-radius:8px;",
+          "font-family:\"Courier New\",monospace;"
+        )
+        girafe_options = list(
+          ggiraph::opts_selection(type = "single"),
```

Comment:
The idea is to have a 4th treeview type which is initially all grey. Then when a user selects a specific mutation from a drop down menu, the nodes which contain that mutation get recoloured to highlight where the mutation occurs.

Initial idea:
- read in (all grey rds file)
- add layer with new coloured points (based on output from ggbuild)
- make widget in app

If I was to add another ggplot layer (before the widget was created), would the `ggiraph::opts_hover_inv()` override those colours?

*I might be able to make a function that just strips the colour out of the existing treeview rds file.

Reply:
An alternative solution would be to use:

-  `session$sendCustomMessage(type = '<output_id>_set', message = selected_nodes)` (to select nodes based on user-selections in a drop-down menu)
- the ggiraph options `opts_selection` (to set the theme for selected nodes)
-  and `opts_selection_inv` (to set the theme for non-selected nodes)

Reply:
The sounds like a better approach than my idea - let's go with that. We don't currently have a table of mutations vs nodes with that mutation. I can manually make a small example for a couple of mutations/nodes to get a working example going, if that would be helpful?

Reply:
A small example is as follows:

```r
# in the UI code
  # allows the user to select nodes based on cluster ID
        shiny::column(
          3,
          shiny::selectInput(ns("nodeChoice"), "Choose a node", c("1601", "1859"))
        )
```

```r
# in the server
    # read in the treeview chart
      g <- reactive({
        filename <- get_filename(input$widgetChoice)
        readRDS(filename)
      })

          observeEvent(input$nodeChoice, {
        # 20 == 1601
        # 25 == 1859

        # the 'node' column contains integers that define the IDs for graph-nodes in the htmlwidget
        node_map <- g()$data[c("cluster_id", "node")]
        selection_map <- dplyr::filter(
          node_map,
          .data[["cluster_id"]] %in% input$nodeChoice
        )

        session$sendCustomMessage(
          # "<output_id>_set" defines the collection of selected nodes on the htmlwidget
          type = paste0(ns("treeview"), "_set"),
          #   but the nodes in the widget are ID's using strings, not integers
          message = as.character(selection_map[["node"]])
        )
      })
```

This is just an example of how to use the custom-messages to highlight a node.

Would the mutation lookup table include all mutations, or just those mutations that are mentioned in the heatmap?